### PR TITLE
Unit Tests for ParticleSpec

### PR DESCRIPTION
**Pull request by alxrsngrtn**
Added more unit tests for ParticleSpec, following this [test plan](http://doc/1AAeLajEGuiDIJ54h2lEkn8LKKOLLN8FaEBTJu-iVf4A).

### CODE REVIEW

File: javatests/arcs/core/data/ParticleSpecTest.kt [hunk 1]
```diff
@@ -8,6 +8,27 @@ import org.junit.runners.JUnit4
...
+
+    assertThat(spec.checks).isEmpty()
+  }
+
+  @Test
+  fun constructor_defaults_noAnnotations() {
+    val spec = ParticleSpec("FooSpec", mapOf(), "path.to.Foo")
+
+    assertThat(spec.annotations).isEmpty()
+  }
```
Comment (alxrsngrtn) [inline javatests/arcs/core/data/ParticleSpecTest.kt:30]: Question for the reviewer: Are these sufficient tests for default constructor args?

File: javatests/arcs/core/data/ParticleSpecTest.kt [hunk 2]
```diff
@@ -48,6 +69,65 @@ class ParticleSpecTest {
     assertFailsWith<IllegalArgumentException> { spec.dataflowType }
   }
 
+  @Test
+  fun egressType_noEgressType_returnsNull() {
```
Comment (mmandlis) [inline javatests/arcs/core/data/ParticleSpecTest.kt:73]: nit - `egressType_noType_returnsNull` name is more consistent with other tests.

File: javatests/arcs/core/data/ParticleSpecTest.kt (hunk 1 above)
Comment (mmandlis) [reply javatests/arcs/core/data/ParticleSpecTest.kt:30]: what about non-default ctor arguments, like connections? should we have tests to test them as well?

File: javatests/arcs/core/data/ParticleSpecTest.kt [hunk 3]
```diff
@@ -48,6 +69,65 @@ class ParticleSpecTest {
     assertFailsWith<IllegalArgumentException> { spec.dataflowType }
```
Comment (mmandlis) [inline javatests/arcs/core/data/ParticleSpecTest.kt:69]: should we have a test case where `spec.dataflowType` doesn't throw?

File: javatests/arcs/core/data/ParticleSpecTest.kt [hunk 4]
```diff
@@ -48,6 +69,65 @@ class ParticleSpecTest {
...
+        Annotation.createEgress(),
+        Annotation.ingress
+      )
+    )
+
+    assertThat(spec.egressType).isNull()
+  }
+
+  @Test
+  fun egressType_dataflowTypeIsIngressAndEgressWithArgument_returnsTypeName() {
```
Comment (mmandlis) [inline javatests/arcs/core/data/ParticleSpecTest.kt:120]: do we need tests for cases when ingressType is specified?

**Review by mmandlis — APPROVED** [review]
overall the tests you added lgtm, I'm wondering, if we should be adding more tests though. feel free to land this and iterate incrementally.

File: javatests/arcs/core/data/ParticleSpecTest.kt (hunk 1 above)
Comment (alxrsngrtn) [reply javatests/arcs/core/data/ParticleSpecTest.kt:30]: It looks like the non-default constructor args aren't used anywhere in the ParticleSpec class. These are data fields in the data class, which in previous discussions, I think we talked about not needing to test.

File: javatests/arcs/core/data/ParticleSpecTest.kt (hunk 3 above)
Comment (alxrsngrtn) [reply javatests/arcs/core/data/ParticleSpecTest.kt:69]: Yes, though it's not shown well in the diff.

File: javatests/arcs/core/data/ParticleSpecTest.kt (hunk 2 above)
Comment (alxrsngrtn) [reply javatests/arcs/core/data/ParticleSpecTest.kt:73]: I like it.

File: javatests/arcs/core/data/ParticleSpecTest.kt (hunk 4 above)
Comment (alxrsngrtn) [reply javatests/arcs/core/data/ParticleSpecTest.kt:120]: 👍

File: javatests/arcs/core/data/ParticleSpecTest.kt (hunk 3 above)
Comment (alxrsngrtn) [reply javatests/arcs/core/data/ParticleSpecTest.kt:69]: (to clarify: there are tests like this that exist).

**Pull request closed without merging**